docs/okeps/okep-5193-user-defined-networks.md

```diff
@@ -0,0 +1,1468 @@
+    //
+    // +optional
+    Lifecycle NetworkIPAMLifecycle `json:"lifecycle,omitempty"`
+}
+
+```
+Suggested API validation rules:
+- `Topology` and `Role` fields are mandatory.
+- `Topology` can be one of `Layer2`, `Layer3`, `Localnet`.
+- `Role` can be one of `Primary`, `Secondary`.
+- `IPAM.Lifecycle` can be `Persistent`.
+- `IPAM.Mode` can be set to `Disabled` only on `Layer2` or `Localnet` topologies for `Secondary` networks, where the `Subnets` parameter must be omitted. When set to `Enabled`, the `Subnets` attribute must be defined.
+- `JoinSubnets` length can be 1 or 2.
+
+#### Cluster scoped CRD
+```golang
+// ClusterUserDefinedNetwork describes shared OVN network across namespaces request.
+type ClusterUserDefinedNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    // +kubebuilder:validation:Required
+    // +required
+    Spec ClusterUserDefinedNetworkSpec `json:"spec"`
+    // +optional
+    Status ClusterUserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// ClusterUserDefinedNetwork defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetwork struct {
+    // NamespaceSelector Label selector for which namespace network should be available for.
+    // +kubebuilder:validation:Required
+    // +required
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector"`
+    
+    // Template is direct specification of UserDefinedNetwork.
+    // +kubebuilder:validation:Required
+    // +required
+    Template *UserDefinedNetworkTemplateSpec `json:"template"`
+}
+
+// UserDefinedNetworkTemplateSpec UserDefinedNetwork spec template.
+type UserDefinedNetworkTemplateSpec struct {
+    // UserDefinedNetworkSpec contains the UserDefinedNetwork specification.
+    Spec UserDefinedNetworkSpec `json:"spec,omitempty"`
+}
+
+// ClusterUserDefinedNetworkStatus contains the observed status of the ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkStatus struct {
+    // ActiveNamespaces indicates in which namespaces network is available.
+    ActiveNamespaces []string `json:"activeNamespaces,omitempty"`
+    
+    Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+Suggested CRD short-name: `cudn`
+
+#### Example
+Condition reflecting the network readiness:
+```yaml
+status:
+  conditions:
+  - type: NetworkReady
+    status: "True"
+    reason: NetworkAttachmentDefinitionReady
+    message: NetworkAttachmentDefinition has been created
+```
+Condition reflecting the network readiness underlying OVN network state:
+```yaml
+status:
+  conditions:
+  - type: NetworkCreated
+    status: "True"
+    reason: OVNNetworkCreated
+    message: OVN network has been created
+```
+
+#### Example - Namespace scoped network
+```yaml
+kind: UserDefinedNetwork
+metadata:
+  name: db-network
+  namespace: demo
+spec:
+  topology: Layer2
+  role: Primary
+  mtu: 9000
+  subnets: ["10.0.0.0/24"]
+  excludeSubnets: ["10.0.0.0/26"]
```

Comment:
Is excludeSubnets actually supported with primary? I don't think we have any tests for this and I remember that in the initial implementation phase - after talking with @qinqon and @maiqueb we decided this won't be supported for primary networks...but only for secondary? -> I am pretty sure we don't exclude this in IPAM right? (or did we update the logical switches to consider this? hmm can't remember)

Reply:
This is supported only for localnet.

Reply:
heh we need it in primary, but yeah will remove for now